pkg/ringbuffer/list.go

```diff
@@ -0,0 +1,64 @@
+package ringbuffer
+
+import (
+	"sync"
+)
+
+var elemPool = &sync.Pool{New: func() interface{} { return new(element) }}
+
+type element struct {
+	next, prev *element
+	val        interface{}
+}
+
+func (e *element) reset() {
+	e.next, e.prev = nil, nil
+	e.val = nil
+}
+
+type queue struct {
+	root  element
+	count int
+}
+
+func newQueue() *queue {
+	q := new(queue)
+
+	q.root.next = &q.root
+	q.root.prev = &q.root
+	q.count = 0
+	return q
+}
+
+func (q *queue) size() int {
+	return q.count
+}
+
+func (q *queue) enqueue(val interface{}) {
+	elem := elemPool.Get().(*element)
+	elem.val = val
+
+	at := q.root.prev
+
+	at.next = elem
+	elem.prev = at
+	elem.next = &q.root
+	q.root.prev = elem
+	q.count++
+}
+
+func (q *queue) dequeue() interface{} {
+	if q.size() == 0 {
+		return nil
+	}
+
+	at := q.root.next
```

Comment:
rename **at** to **first** or **head**, it may more clear

Reply:
Please check the last comment